distributed_shampoo/fsdp_shampoo.py

```diff
@@ -0,0 +1,849 @@
+                )
+            if not dist.get_world_size() % num_trainers_per_group == 0:
+                raise ValueError(
+                    f"Invalid number of GPUs per group: {num_trainers_per_group}. Must divide global world size {dist.get_world_size()}."
+                )
+        else:
+            num_trainers_per_group = 1
+
+        super(FSDPShampoo, self).__init__(
+            params,
+            {
+                LR: lr,
+                BETAS: betas,
+                MOMENTUM: momentum,
+                WEIGHT_DECAY: weight_decay,
+                EPSILON: epsilon,
+                GRAFTING_EPSILON: grafting_epsilon,
+                GRAFTING_BETA2: grafting_beta2,
+            },
+        )
+
+        # Initialize algorithm-related fields.
+        self._max_preconditioner_dim = max_preconditioner_dim
+        self._precondition_frequency = precondition_frequency
+        self._exponent_override = exponent_override
+        self._exponent_multiplier = exponent_multiplier
+        self._num_trainers_per_group = num_trainers_per_group
+        self._use_merge_dims = use_merge_dims
+        self._large_dim_method = large_dim_method
+        self._use_decoupled_weight_decay = use_decoupled_weight_decay
+        self._preconditioner_dtype = preconditioner_dtype
+        self._use_bias_correction = use_bias_correction
+        self._grafting_type = grafting_type
+        self._grafting_epsilon = grafting_epsilon
+        self._grafting_beta2 = grafting_beta2
+        self._parameter_count = 0
+        self._use_nesterov = use_nesterov
+        self._use_protected_eigh = use_protected_eigh
+        self._use_dtensor = use_dtensor
+        self._debug_mode = debug_mode
+        if self._use_nesterov and momentum == 0.0:
+            logger.warning(
+                "Nesterov flag is enabled but momentum parameter is zero! Continuing without using momentum or Nesterov acceleration..."
+            )
+
+        if start_preconditioning_step == -1:
+            self._start_preconditioning_step = precondition_frequency
+            logger.warning(
+                f"start_preconditioning_step set to -1. Setting start_preconditioning_step equal to precondition frequency {precondition_frequency} by default."
+            )
+        elif start_preconditioning_step < precondition_frequency:
+            raise ValueError(
+                f"Invalid start_preconditioning_step value: {start_preconditioning_step}. Must be >= {precondition_frequency = }."
+            )
+        else:
+            self._start_preconditioning_step = start_preconditioning_step
+
+        # Initialize comms-related fields.
+        self._world_size = dist.get_world_size() if self._use_distributed() else 1
+
+        # Initialize Shampoo preconditioners and distributed buffers.
+        self._dist_group = None
+        self._initialize_preconditioners_and_steps()
```

Comment:
Let's remove these variables here since they should not be used.
```suggestion
        self._initialize_preconditioners_and_steps()
```

Reply:
I was thinking that we may need to use them in the future if we decide to experiment with communications, so I left them for now.